### PR TITLE
error messages: simpler names for shadowed identifiers

### PR DESCRIPTION
**Pull request by Octachron**
This PR proposes to use the new naming convention for ambiguous identifiers in the error messages for `illegal include`s (and `generalized open`s).

Currently, when printing identifiers involved in a signature avoidance issue, for instance
```ocaml
include struct
  type t = A
  let x = A
end
type t
```
the error messages resort to using the internal stamp to disambiguate the two types `t`:
```
Error: Illegal shadowing of included type t/1155 by t/1157
       Lines 1-4, characters 0-3:
         Type t/1155 came from this include
       Line 3, characters 6-7:
         The value x has no valid type if t/1155 is shadowed
```
However, for type-level error message, we have recently adopted the convention  that  `t/<n>` means the `n`-th most recent identifier `t` in scope.
This leaves us with two conflicting interpretations for `t/<n>` in error messages.

Moreover, the new convention can be easily applied to `illegal include`s, contrarily to the previous ad-hoc printing mechanism in `Printtyp`: the shadower item should be named `t`  since it is the newest `t` in scope, whereas the shadowed item should be named `t/2`:

```
Error: Illegal shadowing of included type t/2 by t
       File "test.ml", lines 1-4, characters 0-3:
         Type t/2 came from this include
       File "test.ml", line 3, characters 6-7:
         The value x has no valid type if t/2 is shadowed
```

Consequently, I propose to switch to the new convention in both the `illegal include` and `generalized open` error messages in order to decrease the complexity of our error messages.

------------------------------

Note that for generalized opens, 

```ocaml
open struct type t = A end
let x = A
```

the current error message is also using the internal stamp

```
Error: The type t/1160 introduced by this open appears in the signature
       Line 3, characters 4-5:
         The value x has no valid type if t/1160 is hidden
```
but without a second type `t` in the context of the error. Using just `t` in this case seems fine to me.

------------------

Similarly, one might wonder if it might be possible for `illegal include`s to involve more than two types (or components) sharing the same name.

This cannot happen currently because the current typechecker error path reduces signature avoidance issues into a single `illegal include` error that only concerns two components with the same name. 

Unfortunately, this can result in some  `spooky-error-at-distance` phenomenon which are not ideal from my point-of-view, but this is a story for another PR.

(cc @trefis: I remember that we disagreed when we discussed my previous proposition in this area, so if you have any comments or criticisms (or not) on this new proposition, you are very welcome)

### CODE REVIEW

File: testsuite/tests/shadow_include/shadow_all.ml
Comment (Julow) [inline testsuite/tests/shadow_include/shadow_all.ml]: These could be sub-messages with their own source highlighting, like in https://github.com/ocaml/ocaml/pull/11679

Comment (Octachron) [reply testsuite/tests/shadow_include/shadow_all.ml]: Good idea, but I think it is probably better to discuss this change of format once I will have submitted my second PR in this series which handle the case where there is a set of illegal includes but no single include is illegal in isolation.

**Comment by gasche** [comment]
@Julow I am assuming that you are "taking care" of this PR and planning to do a full review, possibly approve if you like the change. (We would probably wait for mid-February before considering a merge anyway.) Let me know if I am overly optimistic about this.

File: typing/printtyp.ml
Comment (Julow) [inline typing/printtyp.ml]: This explanation could be moved to the interface as a docstring.

File: typing/typemod.ml
Comment (Julow) [inline typing/typemod.ml]: The hardcoded number won't be ambiguous to the user because the error message explain where it comes from. But could an other error message bind `t/2` to something else ?
`Printtyp.ident` seems to be more careful at picking an index. Perhaps it could be used here with `shadowed_item_id` added to the Env ?

**Review by Julow — COMMENTED** [review]
Thanks @gasche !

File: typing/printtyp.ml
Comment (Octachron) [reply typing/printtyp.ml]: Indeed, thanks for the suggestion!

File: typing/typemod.ml
Comment (Octachron) [reply typing/typemod.ml]: Each error message comes with its own context and `Env.t`, which is signaled to the user through the location of the error message. In other words, a name like `t/<n>` is only bound inside the context of a specific error message.

Moreover those indices are the indices that we would get if we computed the environment at the end of the definition of the shadower item, excluding ephemeral item definitions: at the end of the definition of the shadower item, it is the most recent item with its name. Then, the shadowed item is necessarily the previous non-ephemeral component sharing this name: if there were another non-ephemeral component between the shadower and shadowed itself, we would have reported this definition has an illegal shadowing before the current error.

Nevertheless, I propose to amend that behavior to also count ephemeral components:
```
include struct type t = A end
let x  = A
open struct type t = int end
type t
```
currently refers to the shadowed `t` as `t/2` when we probably want to name `t/3` which is done in #11916 in https://github.com/ocaml/ocaml/pull/11916/commits/f47ea0ccfd55af7532c216c3669f43991954e2c8 but I didn't propagate the change because that I forgot the issue with intermediary ephemeral component definitions.

Comment (Julow) [reply typing/typemod.ml]: Thanks for the explanation! Would it make sense to make `Printtyp.ident` aware of the `erased_idents` of https://github.com/ocaml/ocaml/commit/f47ea0ccfd55af7532c216c3669f43991954e2c8 and use that instead of `Printtyp.human_id` ?
It seems that it was supposed to handle such cases without asking for an arbitrary index. `human_id` could be left un-exposed.

Comment (Octachron) [reply typing/typemod.ml]: The issue of using `Printtyp.ident` directly is that for `single illegal include`s is that we don't have an environment at hand at the point of the shadower identifier. We could re-add the erased identifiers to the printing environment of the error but I am not sure if building an artificial environment is more readable than building a specialized identifier function. It might be the case than the environement is the best option because it require less knowledge of the printing machinery of `Pprintyp`, I will think about it.

File: testsuite/tests/shadow_include/shadow_all.ml
Comment (Octachron) [reply testsuite/tests/shadow_include/shadow_all.ml]: I couldn't resist to the temptation of switching to sub-messages since the implementation is straightforward and I find the resulting error message quite better.

File: typing/typemod.ml
Comment (Octachron) [reply typing/typemod.ml]: After more investigations, using a `Printtyp.namespaced_ident` function with the error environment works well enough once the "usual" interaction between the global state of `Printtyp` and the delayed printing of `Location.msg` is correctly handled.

File: typing/printtyp.ml
Comment (Julow) [inline typing/printtyp.ml:320]: This might easier to read if it was a `when`:
```
  | Some namespace when !enabled && not (fuzzy_id namespace id) ->
```

**Review by Julow — APPROVED** [review]
Looks good to me :)

**Review by gasche — APPROVED** [review]
Approved mostly on @Julow's behalf.

**Pull request merged**
